### DISCUSSION TITLE
Store user Input data from R Shiny to Google Sheets

### DISCUSSION THREAD

**Discussion by hendrarespati** [teamalgoritma/community#429] [Q&A] [answered]
Hi Guys,

currently i'm in project to create web app forms using Rshiny and want to store the input data to google sheet.. i try to follow this tutorial: https://debruine.github.io/shinyintro/data.html 

but always return message. 

Error in `gargle::response_process()`:
! Client error: (403) PERMISSION_DENIED
• Client does not have sufficient permission. This can happen because the OAuth token does not have
  the right scopes, the client doesn't have permission, or the API has not been enabled for the
  client project.
• Request had insufficient authentication scopes.

anyone can help me about tutorial to store our data from rshiny to google sheet?

sorry i cannot share the codes because its my Office Project..

P.S: I'm using OS: Linux Ubuntu  18.04.6 LTS Bionic in Virtual machine Google cloud

Many Thanks,

Hendra Respati
Batch 1 Algrotima

**Comment by finesaaa** [comment, marked as answer]
Dear Pak Hendra, 

Berikut kami terdapat contoh lain untuk tutorial integrasi Shiny dengan Google Sheet, boleh Bapak coba terlebih dahulu untuk data Bapak. Jika masih belum bisa, mohon berkabar disini.
https://medium.com/@predict42/google-sheets-in-r-shiny-a-happy-marriage-8812303711ac

**Reply by hendrarespati** [reply]
Hi Finesaaa,

Thanks untuk referensi linknya.. sy sudah googling itu dan dah mencoba namun gagal. kenapa? karena tutorial tersebut dah outdated, package "googlesheet dah ga dipakai di versi R 3.5.0 keatas'dan telah diganti menjadi "googlesheet4". 

namun sy sudah menemukan how to connect rshiny to Google sheet & Google drive.
bagaimana caranya?

1. create service account di google cloud di Api & Services > Credentials > Create Credentials > Service Accopunt.

2. salin email yang di generate automatis dari service account yg telah dibuat : <your id>@<usually your project name in Gcloid>.iam.gserviceaccount.com, contoh: rshiny@test_project.iam.gserviceaccount.com

3. pada file google spreadsheet yg akan di akses atau di edit pastikan untuk memberikan akses ke email yg digenerate dr service account tadi. beri role as editor untuk email tersebut.

4. download file json dr service account yg telah dibuat. isi file json ini berisi seluruh informasi yg diperlukan untuk proses authentification nanti. cth: test_project-7e41184b65a9,json

5. taruh file json tersebut di folder project rshiny dibuat untuk kasus saya letaknya di: "/home/hendra/app/form_webapp/test_project-7e41184b65a9.json" 

6. load package googledrive & googlesheets4 dan tulis code tuk authentification ke google drive dan google sheet:

library(googledrive)
library(googlesheets4)

drive_auth(path = "/path/for/your/json file/stored/your_json_file_name.json")
gs4_auth(path = ""/path/for/your/json file/stored/your_json_file_name.json"")


Sekarang anda sudah bs tuk membaca dan mengedit spreadsheet di google sheet


best regards,
Hendra Respati
Batch 1 Algoritma

**Reply by finesaaa** [reply]
Terima kasih kembali untuk sharing ilmunya pak! Saya beberapa hari juga langsung mencoba googlesheet4, yang saya cob bisa dicek di [Artikel Medium Saya](https://fiqey.medium.com/using-googlesheets-as-a-database-in-r-happy-marriage-372cd4fc4770). Senang berbagi ilmu dengan Bapak Hendra!
